### PR TITLE
Added edge_intelligence examples and edge_component packages along with tests.

### PR DESCRIPTION
**Pull request by lucifercr07**
1. Added ModelPath to conf file for the edge_intelligence_models e.g windmill_iotcc_pfa.py etc.
2. Removed the training files from edge_intelligence_models i.e eval-build.py and build-model.py
3. Added readme information to the edge_intelligence_models.
4. Removed file_reader_package.py from edge_component
5. Added comments for actuator in edge_intelligence_models e.g windmill_iotcc_pfa.py etc.
6. Added loadModel for the tensor_flow_edge_component.py
7. Added files windmill_graphite_tf.py and windmill_iotcc_tf.py to the edge_intelligence_models.
8.  Added rule_edge_component
9. Added windmill_graphite_rule.py 
10. Added packages for edge_components (sklearn_edge_component, pfa_component, 
     tensorflow_edge_component, rule_edge_component.py).
11. Example packages for each edge_component (graphite_pfa_windmill_simulated.py, 
     graphite_rule_thermistor_simulated.py, graphite_rule_windmill_simulated.py, 
     graphite_sklearn_windmill_simulated.py, graphite_tensorflow_windmill_simulated.py).
12. Tests for edge_components under tests folder.

### CODE REVIEW

File: examples/sampleProp.conf [hunk 1]
```diff
@@ -20,3 +20,7 @@ IotCCPassword = "Password"
 
 GraphiteIP = "Graphite-IP"
 GraphitePort = None
+
+#### [EDGE_INTELLIGENCE_MODEL_PATH] ####
+
+ModelPath = "Model-Path"
```
Comment (KohliDev) [inline examples/sampleProp.conf]: Add a new line here.

**Review by KohliDev — APPROVED** [review]

File: examples/sampleProp.conf (hunk 1 above)
Comment (lucifercr07) [reply examples/sampleProp.conf]: I have done the change.

File: edge_intelligence_models/README.md [hunk 2]
```diff
@@ -1,3 +1,5 @@
 # Edge Intelligence Models
 
-Place holder for Edge intelligence models
\ No newline at end of file
+Saved pfa, sklearn and tensorflow models are present under the windmill-models directory.
```
Comment (KohliDev) [inline edge_intelligence_models/README.md]: Trained pfa ...

<PFA, SKlearn , Tensorflow link>
The generated data is using the TI sensor tag which has been used for training the windmill model.
It provides a basic example which uses TI sensor tag <link> data to simulate a sample windmill model. The basic metrics we collect from Windmill model are:

File: edge_intelligence_models/windmill-models/saved_tf_model/checkpoint [hunk 3]
```diff
@@ -0,0 +1,3 @@
+model_checkpoint_path: "model.ckpt-2000"
```
Comment (KohliDev) [inline edge_intelligence_models/windmill-models/saved_tf_model/checkpoint]: Trained_tf_model

File: edge_intelligence_models/windmill_tf_model.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+from __future__ import absolute_import
```
Comment (KohliDev) [inline edge_intelligence_models/windmill_tf_model.py]: Remove python3 support, add the license.
Move this to example/edge_intelligence/trained_models

File: edge_intelligence_models/eval-model.py [hunk 5]
```diff
@@ -1,90 +0,0 @@
-import tensorflow as tf
```
Comment (KohliDev) [inline edge_intelligence_models/eval-model.py:1]: Change file permission to 644

File: edge_intelligence_models/windmill_test.csv [hunk 6]
```diff
@@ -0,0 +1,31 @@
+30,1
```
Comment (KohliDev) [inline edge_intelligence_models/windmill_test.csv]: Explain the usage of all files

File: edge_intelligence_models/windmill_tf_model.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import os
+import urllib
+
+import numpy as np
+import tensorflow as tf
+
+# Data sets
```
Comment (KohliDev) [inline edge_intelligence_models/windmill_tf_model.py]: Add comments.

File: edge_intelligence_models/windmill_tf_model.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      filename=TEST_SET,
+      target_dtype=np.int,
+      features_dtype=np.float32)
+
+  # Specify that all features have real-value data
+  feature_columns = [tf.contrib.layers.real_valued_column("", dimension=1)]
+
+  
+  classifier = tf.contrib.learn.LinearClassifier(feature_columns=feature_columns,
+                                              model_dir="/home/prasha/git/edge_intelligence/edge_intelligence_models/windmill-model/saved_tf_model")
```
Comment (KohliDev) [inline edge_intelligence_models/windmill_tf_model.py]: trained_tf_model/

File: examples/edge_intelligence/Readme.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
+This folder contains example files for different models: pfa, sklearn, tensorflow, rule.
```
Comment (KohliDev) [inline examples/edge_intelligence/Readme.md]: PFA, RuleEdge

File: examples/edge_intelligence/Readme.md [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+# RuleEdgeComponent
+
+Rule edge component is like different edge components provided by Liota, you can define a simple rule depending on your requirement as a lambda function and the rule edge component will process the incoming metrics and give output to the actuator on the basis of the rule you have defined as lambda function.
```
Comment (KohliDev) [inline examples/edge_intelligence/Readme.md]: one of the edge intelligences components.

Lambda <Link Python lambda>

The result can be passed onto actuator on basis of the rule you have defined as lambda function.

File: examples/edge_intelligence/Readme.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+# Example
+If you have to take rpm as a metric and if rpm exceeds certain limits you have to perform some action, you can define a lambda function as:
+```
+Rule = lambda x : 0 if (rpm>=rpm_limit) else 0
+```
+Now just pass the rule to RuleEdgeComponent along with the action_actuator and exceed_limit.
```
Comment (KohliDev) [inline examples/edge_intelligence/Readme.md]: Create the ruleEdge and pass it onto RuleEdgeComponent.

Action_actuator_parameter explain
exceed_limit explin
actuator_udm explain with full form

File: examples/edge_intelligence/Readme.md [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+exceed_limit is a parameter which can be specified about after how many times the limit is exceeded then action should be taken.
+example: if rpm exceeds rpm_limit consecutively 3 times then only action should be applied, in that case we would assign exceed_limit = 3
+```
+edge_component = RuleEdgeComponent(ModelRule,exceed_limit,actuator_udm=action_actuator)
+```
+After this just register the edge_component and start_collecting the values from sensor.
```
Comment (KohliDev) [inline examples/edge_intelligence/Readme.md]: the values which will get published to RuleEdge component. Please refer the example.

File: examples/sampleProp.conf [hunk 13]
```diff
@@ -1,6 +1,6 @@
 #### [DEFAULT] ####
 
-EdgeSystemName = "EdgeSystem-Name"
+EdgeSystemName = "EdgeSystem-prasha"
```
Comment (KohliDev) [inline examples/sampleProp.conf]: Remove all the parameters value

File: examples/edge_intelligence/windmill_graphite_pfa.py [hunk 14]
```diff
@@ -191,7 +191,7 @@ def action_actuator(value):
         temperature_metric = Metric(
             name="windmill.AmbientTemperature",
             unit=ureg.degC,
-            interval=0,
+            interval=10,
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_pfa.py]: Check in SensorTag before committing this code.

File: examples/edge_intelligence/windmill_graphite_pfa.py [hunk 15]
```diff
@@ -261,10 +261,9 @@ def action_actuator(value):
             aggregation_size=1,
             sampling_function=get_rpm_for_model
         )
-
-        edge_component = PFAComponent(
-            '/home/pi/Desktop/Borathon/liota/edge_intelligence_models/windmill/finalized_model.pfa',
-            actuator_udm=action_actuator)
+        #Model-Path can be edited in the sampleProp.conf file
+        #pass value to actuator as of now the action_actuator prints the value on the console
+        edge_component = PFAComponent(config['ModelPath'], actuator_udm=action_actuator)
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_pfa.py]: actuator_udm and mention the comment before function.

File: examples/edge_intelligence/windmill_graphite_rule.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+from liota.dccs.graphite import Graphite
+from liota.dcc_comms.socket_comms import SocketDccComms 
+from liota.entities.metrics.metric import Metric 
+from liota.dccs.dcc import RegistrationFailure
+from liota.edge_component.rule_edge_component import RuleEdgeComponent 
+from liota.entities.edge_systems.dell5k_edge_system  import Dell5KEdgeSystem
+import random
+
+config = {}
+execfile('../sampleProp.conf', config)
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_rule.py:43]: Path correct

File: examples/edge_intelligence/windmill_graphite_rule.py [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+	return random.randint(42,54)
+
+def get_vibration():
+	return round(random.uniform(0.480,0.7),3)
+
+def get_temp():
+	return round(random.uniform(2.0,7.0),2)
+
+def collect_rpm_vib():
+	l = [get_rpm(), get_vibration()]
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_rule.py]: rename it two parameters mention in READMe currently tested for single metric.

File: examples/edge_intelligence/windmill_graphite_rule.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+def collect_rpm_vib_temp():
+	l = [get_rpm(), get_vibration(), get_temp()]
+	return l
+
+def action_actuator(value):
+	print value
+
+if __name__ == '__main__':
+
+	graphite = Graphite(SocketDccComms(ip=config['GraphiteIP'],port=8080))
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_rule.py]: 8080 read from sampleProp.conf

File: examples/edge_intelligence/windmill_graphite_rule.py [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+		rule_rpm_metric = Metric(
+			name="windmill.RPM",
+			unit=None,
+			interval=1,
+			aggregation_size=1,
+			sampling_function=get_rpm
+		)
+		
+		rpm_limit=45
+		ModelRule = lambda x : 1 if (x>=rpm_limit) else 0
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_rule.py]: x rename as value_Collected

Mention  a comment describing the model

File: examples/edge_intelligence/windmill_graphite_sklearn.py [hunk 20]
```diff
@@ -55,72 +55,72 @@
 
 
 def read_cpu_procs():
-    return cpu_stat.procs_running()
+	return cpu_stat.procs_running()
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_sklearn.py]: Formatting of the code

File: examples/edge_intelligence/windmill_graphite_sklearn.py [hunk 21]
```diff
@@ -55,72 +55,72 @@
...
 
 
 def get_rpm_for_model():
-    return rpm_model_queue.get(block=True)
+	return rpm_model_queue.get(block=True)
 
 
 def action_actuator(value):
-    print value
+	print value
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_sklearn.py]: actuator_udm can be used to pass on the value to the actuator, as of now we are printing them

File: examples/edge_intelligence/windmill_graphite_sklearn.py [hunk 22]
```diff
@@ -130,145 +130,144 @@ def action_actuator(value):
...
+		mem_free_metric = Metric(
+			name="windmill.Memory_Free",
+			unit=None,
+			interval=10,
+			sampling_function=read_mem_free
+		)
+		reg_mem_free_metric = graphite.register(mem_free_metric)
+		graphite.create_relationship(reg_edge_system, reg_mem_free_metric)
+		reg_mem_free_metric.start_collecting()
+		'''
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_sklearn.py]: Remove this

File: examples/edge_intelligence/windmill_graphite_sklearn.py [hunk 23]
```diff
@@ -130,145 +130,144 @@ def action_actuator(value):
...
+		# Connects to the SensorTag device over BLE
+		sensor_tag_collector = SensorTagCollector(device_name=config['DeviceName'], device_mac=config['DeviceMac'],
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_sklearn.py]: Chane sensor tag implementation as per the PR

https://github.com/vmware/liota/pull/90

File: examples/edge_intelligence/windmill_graphite_sklearn.py [hunk 24]
```diff
@@ -130,145 +130,144 @@ def action_actuator(value):
...
+		reg_rpm_metric.start_collecting()
+		'''
+		sklearn_rpm_metric = Metric(
+			name="windmill.RPM",
+			unit=None,
+			interval=0,
+			aggregation_size=1,
+			sampling_function=get_rpm_for_model
+		)
+		#Model-Path can be edited in the sampleProp.conf file
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_sklearn.py]: Comment formatting.

File: examples/edge_intelligence/windmill_graphite_sklearn.py [hunk 25]
```diff
@@ -130,145 +130,144 @@ def action_actuator(value):
...
+		#pass value to actuator as of now the action_actuator prints the value on the console
+		edge_component = SKLearnEdgeComponent(config['ModelPath'], actuator_udm=action_actuator)
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_sklearn.py]: Edge Component comments

File: examples/edge_intelligence/windmill_graphite_tf.py [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+            name="windmill.RPM",
+            unit=None,
+            interval=1,
+            aggregation_size=1,
+            sampling_function=get_rpm
+        )
+
+        #Model-Path can be edited in the sampleProp.conf file
+        #pass value to actuator as of now the action_actuator prints the value on the console
+        edge_component = TensorFlowEdgeComponent(config['ModelPath'], actuator_udm=action_actuator)
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_graphite_tf.py]: Mention the comment TensorFlor model

File: examples/edge_intelligence/windmill_iotcc_pfa.py [hunk 27]
```diff
@@ -266,9 +266,10 @@ def action_actuator(value):
             aggregation_size=1,
             sampling_function=get_rpm_for_model
         )
-
-        edge_component = PFAComponent('/home/pi/Desktop/Borathon/liota/edge_intelligence_models/windmill/finalized_model.pfa',
-                                      actuator_udm=action_actuator)
+        
+        #Model-Path can be edited in the sampleProp.conf file
+        #pass value to actuator as of now the action_actuator prints the value on the console
+        edge_component = PFAComponent(config['ModelPath'],actuator_udm=action_actuator)
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_iotcc_pfa.py]: Comment about PFA component.

Comment (KohliDev) [reply examples/edge_intelligence/windmill_iotcc_pfa.py]: Show one metric publishing to DCC same metric to Edge_Component

File: examples/edge_intelligence/windmill_iotcc_sklearn.py [hunk 28]
```diff
@@ -259,19 +259,19 @@ def action_actuator(value):
...
+        sklearn_rpm_metric = Metric(
             name="RPM",
             unit=None,
             interval=0,
             aggregation_size=1,
             sampling_function=get_rpm_for_model
         )
+        #Model-Path can be edited in the sampleProp.conf file
+        #pass value to actuator as of now the action_actuator prints the value on the console
+        edge_component = SKLearnEdgeComponent(config['ModelPath'],actuator_udm=action_actuator)
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_iotcc_sklearn.py]: Mention comment about Sklearn

File: examples/edge_intelligence/windmill_iotcc_tf.py [hunk 29]
```diff
@@ -0,0 +1,277 @@
...
+
+        tf_rpm_metric = Metric(
+            name="RPM",
+            unit=None,
+            interval=0,
+            aggregation_size=1,
+            sampling_function=get_rpm_for_model
+        )
+
+        #Model-Path can be edited in the sampleProp.conf file
```
Comment (KohliDev) [inline examples/edge_intelligence/windmill_iotcc_tf.py]: Add comment and show one metric, correct the comment formatting.

File: examples/sampleProp.conf [hunk 30]
```diff
@@ -11,12 +11,20 @@ SystemPropList = {"Country":"USA-G", "State":"California", "City":"Palo Alto", "
...
 IotCCPassword = "Password"
 
 #### [GRAPHITE] ####
 
-GraphiteIP = "Graphite-IP"
-GraphitePort = None
+GraphiteIP = "92.246.246.188"
+GraphitePort = 2003
+
+#### [EDGE_INTELLIGENCE_MODEL_PATH] ####
```
Comment (KohliDev) [inline examples/sampleProp.conf]: Edge_Intelligence

File: setup.py [hunk 31]
```diff
@@ -157,4 +157,4 @@
                 (os.path.abspath(os.sep) + '/../etc/liota',
                  ['BSD_LICENSE.txt', 'BSD_NOTICE.txt']),
                 (os.path.abspath(os.sep) + '/../var/log/liota', [])]
-)
+)
```
Comment (KohliDev) [inline setup.py]: Add extra line

File: examples/edge_intelligence/windmill_graphite_rule.py (hunk 16 above)
Comment (lucifercr07) [reply examples/edge_intelligence/windmill_graphite_rule.py:43]: It's correct checked.

File: edge_intelligence_models/eval-model.py (hunk 5 above)
Comment (lucifercr07) [reply edge_intelligence_models/eval-model.py:1]: All files which are being shown outdated moved to examples/edge_intelligence/trained_models
